File: sdk/typescript/src/utils/cache.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 KirkyX. All rights reserved.

/**
 * Response Cache Utility
 *
 * Provides simple in-memory caching for API responses to reduce redundant requests.
 */

/**
 * A cache entry with value and expiration.
 */
interface CacheEntry<T> {
  value: T;
  expiresAt: number;
}

/**
 * Performance metrics for the cache.
 */
export interface CacheMetrics {
  hits: number;
  misses: number;
  entries: number;
  cleanupCount: number;
}

/**
 * Simple in-memory response cache with TTL support.
 */
export class ResponseCache<T = unknown> {
  private readonly cache: Map<string, CacheEntry<T>>;
  private readonly defaultTtl: number;
  private readonly maxEntries: number;
  private metrics: CacheMetrics;

  /**
   * Create a new response cache.
   *
   * @param defaultTtl Default time-to-live in seconds
   * @param maxEntries Maximum number of entries (default: 1000)
   */
  constructor(defaultTtl: number = 60, maxEntries: number = 1000) {
    this.cache = new Map();
    this.defaultTtl = defaultTtl;
    this.maxEntries = maxEntries;
    this.metrics = {
      hits: 0,
      misses: 0,
      entries: 0,
      cleanupCount: 0,
    };
  }

  /**
   * Get a value from cache.
   *
   * @param key Cache key
   * @returns Cached value if exists and not expired, null otherwise
   */
  get(key: string): T | null {
    const entry = this.cache.get(key);
    if (!entry) {
      this.metrics.misses++;
      return null;
    }

    if (entry.expiresAt < Date.now()) {
      // Expired, remove it
      this.cache.delete(key);
      this.metrics.entries--;
      this.metrics.misses++;
      return null;
    }

    this.metrics.hits++;
    return entry.value;
  }

  /**
   * Set a value in cache.
   *
   * @param key Cache key
   * @param value Value to cache
   * @param ttl Time-to-live in seconds (uses default if not specified)
   */
  set(key: string, value: T, ttl?: number): void {
    const entry: CacheEntry<T> = {
      value,
      expiresAt: Date.now() + (ttl ?? this.defaultTtl) * 1000,
    };

    // Enforce max entries limit
    if (this.cache.size >= this.maxEntries && !this.cache.has(key)) {
      // Remove oldest entry (simple FIFO)
      const firstKey = this.cache.keys().next().value;
      if (firstKey) {
        this.cache.delete(firstKey);
        this.metrics.entries--;
      }
    }

    this.cache.set(key, entry);
    this.metrics.entries++;
  }

  /**
   * Delete a value from cache.
   *
   * @param key Cache key
   * @returns true if key was deleted, false if not found
   */
  delete(key: string): boolean {
    const deleted = this.cache.delete(key);
    if (deleted) {
      this.metrics.entries--;
    }
    return deleted;
  }

  /**
   * Clear all cache entries.
   */
  clear(): void {
    this.cache.clear();
    this.metrics.entries = 0;
  }

  /**
   * Remove all expired entries.
   *
   * @returns Number of entries removed
   */
  cleanupExpired(): number {
    const now = Date.now();
    const expiredKeys: string[] = [];

    for (const [key, entry] of this.cache.entries()) {
      if (entry.expiresAt < now) {
        expiredKeys.push(key);
      }
    }

    for (const key of expiredKeys) {
      this.cache.delete(key);
      this.metrics.entries--;
    }

    this.metrics.cleanupCount += expiredKeys.length;
    return expiredKeys.length;
  }

  /**
   * Get the number of entries in cache.
   *
   * @returns Number of cache entries
   */
  size(): number {
    return this.cache.size;
  }

  /**
   * Get cache metrics.
   *
   * @returns Cache performance metrics
   */
  getMetrics(): CacheMetrics {
    return { ...this.metrics };
  }

  /**
   * Reset metrics.
   */
  resetMetrics(): void {
    this.metrics = {
      hits: 0,
      misses: 0,
      entries: this.cache.size,
      cleanupCount: 0,
    };
  }

  /**
   * Get cache hit rate.
   *
   * @returns Hit rate as percentage (0-100)
   */
  getHitRate(): number {
    const total = this.metrics.hits + this.metrics.misses;
    return total > 0 ? (this.metrics.hits / total) * 100 : 0;
  }
}

File: sdk/rust/src/utils/cache.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 KirkyX. All rights reserved.

use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use std::time::{Duration, Instant};

#[derive(Debug, Clone)]
struct CacheEntry<T> {
    value: T,
    expires_at: Instant,
}

#[derive(Debug, Clone, Default)]
pub struct CacheMetrics {
    pub hits: u64,
    pub misses: u64,
    pub entries: u64,
    pub cleanup_count: u64,
}

pub struct ResponseCache<T> {
    cache: Arc<RwLock<HashMap<String, CacheEntry<T>>>>,
    default_ttl: Duration,
    max_entries: usize,
    metrics: Arc<RwLock<CacheMetrics>>,
}

impl<T: Clone> ResponseCache<T> {
    pub fn new(default_ttl: Duration, max_entries: usize) -> Self {
        Self {
            cache: Arc::new(RwLock::new(HashMap::new())),
            default_ttl,
            max_entries,
            metrics: Arc::new(RwLock::new(CacheMetrics::default())),
        }
    }

    pub fn default() -> Self {
        Self::new(Duration::from_secs(60), 1000)
    }

    /// Returns cloned value for safety (avoids lifetime issues with locked data)
    pub fn get(&self, key: &str) -> Option<T> {
        let cache = self.cache.read().unwrap();

        if let Some(entry) = cache.get(key) {
            if entry.expires_at > Instant::now() {
                let mut metrics = self.metrics.write().unwrap();
                metrics.hits += 1;
                return Some(entry.value.clone());
            } else {
                // Entry expired - need to acquire write lock to remove
                drop(cache); // Release read lock before acquiring write lock
                let mut cache = self.cache.write().unwrap();
                cache.remove(key);
                let mut metrics = self.metrics.write().unwrap();
                metrics.entries = cache.len() as u64;
                metrics.misses += 1;
                return None;
            }
        }

        let mut metrics = self.metrics.write().unwrap();
        metrics.misses += 1;
        None
    }

    pub fn set(&self, key: String, value: T, ttl: Option<Duration>) {
        let mut cache = self.cache.write().unwrap();
        let mut metrics = self.metrics.write().unwrap();

        if cache.len() >= self.max_entries && !cache.contains_key(&key) {
            // Remove oldest entries (simple strategy: remove first 10%)
            let keys_to_remove: Vec<String> = cache.keys().take(self.max_entries / 10).cloned().collect();
            for k in keys_to_remove {
                cache.remove(&k);
            }
        }

        let expires_at = Instant::now() + ttl.unwrap_or(self.default_ttl);
        cache.insert(key, CacheEntry { value, expires_at });
        metrics.entries = cache.len() as u64;
    }

    pub fn delete(&self, key: &str) {
        let mut cache = self.cache.write().unwrap();
        let mut metrics = self.metrics.write().unwrap();
        cache.remove(key);
        metrics.entries = cache.len() as u64;
    }

    pub fn clear(&self) {
        let mut cache = self.cache.write().unwrap();
        let mut metrics = self.metrics.write().unwrap();
        cache.clear();
        metrics.entries = 0;
    }

    pub fn cleanup_expired(&self) -> usize {
        let mut cache = self.cache.write().unwrap();
        let mut metrics = self.metrics.write().unwrap();

        let now = Instant::now();
        let mut removed = 0;

        cache.retain(|_, entry| {
            if entry.expires_at > now {
                true
            } else {
                removed += 1;
                false
            }
        });

        metrics.entries = cache.len() as u64;
        metrics.cleanup_count += 1;
        removed
    }

    pub fn size(&self) -> usize {
        let cache = self.cache.read().unwrap();
        cache.len()
    }

    pub fn get_metrics(&self) -> CacheMetrics {
        let metrics = self.metrics.read().unwrap();
        metrics.clone()
    }

    pub fn reset_metrics(&self) {
        let mut metrics = self.metrics.write().unwrap();
        *metrics = CacheMetrics::default();
    }

    pub fn get_hit_rate(&self) -> f64 {
        let metrics = self.metrics.read().unwrap();
        let total = metrics.hits + metrics.misses;
        if total == 0 {
            0.0
        } else {
            metrics.hits as f64 / total as f64
        }
    }
}